File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        const int N = nums.size();
        
        return DFS(nums, 0, N);
    }
    
private:
    bool DFS(vector<int>& nums, int currIdx, const int N) {
        
        if(currIdx >= N || nums[currIdx] == -1) return false;
        if(currIdx == N -1) return true;
        
        int x = nums[currIdx];
        nums[currIdx] = -1;
        
        for(int i = 1; i <= x; i++) {
            if(DFS(nums, currIdx +i, N)) return true;
        }
        
        return false;
    }
};

File: 0055-jump-game/0055-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        final int N = nums.length;
        int maxPosCanGo = 0;
        
        for(int i = 0; i < N; i++) {
            maxPosCanGo = Math.max(maxPosCanGo, i +nums[i]);
            if(maxPosCanGo == i) break;
            
        }
        
        return (maxPosCanGo >= N -1);
    }
}